### ISSUE TITLE
[0.9.1.0-ish] Multi-occupant elections suffer from spoiler effect

### ISSUE THREAD

**Issue by ThePiachu** [StrangeLoopGames/EcoIssues#19204]
So a few days back we had an election on White Tiger:

http://gs1.play.eco:3011/elections.html?election=27

As far as I remember, it went like this - Hindenburg and tkaloger applied as candidates first, Hindenburg got a lot of first votes, tkaloger got a few second place votes and a few first picks. However, I decided to run as well a bit later and got the same amount of first votes. The result of this election was that Hindenburg got in but the two other candidates tied and neither got in. However, this seems to be a spoiler effect problem in politics - people voting for the most popular candidate made other ones not get their votes.

Basically, Out of the 14 people that voted for Hindenburg first, more would've voted for tkaloger than me, yet those didn't count, despite the elections being for 2 seats instead of one. Here is a convenient CGP Grey video on the matter:

https://youtu.be/l8XOZJkozfI?list=PLNCHVwtpeBY4mybPkHEnRxSOb7FQ2vF9c&t=249

Basically, hindenburg needed ~9 votes to win, so their remaining 5 votes should've been distributed based on their voter's second choice to get the second winner.

This isn't a problem for electing officials with only one seat, but is a problem for multi-seat positions.

**Comment by tomsseisums** [comment]
From estimation meeting: it would be good, if someone could describe in more detail, as the voting calculation logic is not immediately clear for development.

**Comment by ThePiachu** [comment]
Okay, I think this is how you can resolve the voting unambiguosly:

Assumptions:
- Every vote cast has a weight that starts at 1, and you have a list of preferred candidates as usual.
- Candidates are ranked first by number of votes descending, and the date they applied to the election ascending (candidate that applied last wins ties, since they got the most amount of "votes per hour" so to say)

Then the algorithm can go like this:

1) If number of candidates <= Number of Seats to Fill, they all get elected, END. If no more Seats to Fill, END.
2) Tally all the first choice votes, rank the candidates
3) Does the top candidate have more votes than (Total Votes Cast / Number of Seats to Fill)?
4) If YES, they get elected. Take all votes that voted for them, multiply their voting power by (1 - (Total Votes Cast / (Votes Cast for Them * Number of Seats to Fill))). Remove them from ALL the votes (so second choices become first choices, etc.). GOTO 1
5) If NO, take the lowest ranking candidate (the one with fewest votes first, and the one that joined the election earliest second), remove them from ALL the votes. GOTO 1

You could try breaking an election down to electing 1 person and then running a sub-election after Step 4 with the adjusted voting weight and 1 less Seat to Fill, but it might not be necessary.

Adjusting the weight of the votes ensures that the spoiler effect doesn't hinder the second choices. So if someone had 9 out of 10 votes for a 2 position election, their secondary votes will be worth (1 - (10 / (9*2))) = (1-(5/9)) = (4/9) weight each, or a total of 4 votes worth for all the remaining 9 votes (aka, 4 votes over the 5 needed to win one seat).

Then, ordering the candidates by the shortest time campaigning ensures there are no ties in the election. Since ECO elections allow candidates to enter and be voted on at any time, rather than the candidates being "locked in" before the votes start, it's only fair to give someone that joined later that small leg up.

Finally, some votes might end up being blank after all the ranked candidates have been eliminated, so there is the fallback of "if number of seats <= number of candidates, elect everyone" to fill all the seats if possible even if all the remaining votes are blank.

I hope this helps!
